Optimized built-in factory lookup

YohDeadfall opened the pull request:
While there are only two built-in factories, there is no need to insatiate a list which complicates enumeration.

YohDeadfall committed:
Optimized built-in factory lookup

stephentoub commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:
```diff
@@ -40,19 +40,20 @@ public sealed partial class JsonSerializerOptions
             // We will likely add collection converters here in the future.
 
-            Debug.Assert(NumberOfConverters == converters.Count);
+            Debug.Assert(NumberOfConverters == converters.Length);
```
NumberOfConverters can be deleted, along with this assert.  In fact, this whole method can be removed, leaving just:
```C#
private static readonly JsonConverter[] s_defaultFactoryConverters = new JsonConverter[]
{
    new JsonConverterEnum(),
    new JsonKeyValuePairConverter()
};
```

YohDeadfall replied:
Will do, but could you explain the idea of the previous implementation? It was possible to do the same thing using an in-place list initializer.

stephentoub replied:
> It was possible to do the same thing using an in-place list initializer.

Not exactly.  I didn't write it, but presumably the author wanted to ensure that the list was being correctly sized, such that it would allocate a backing array of exactly the required size and not need to grow, e.g. ensuring that if someone accidentally added:
```C#
converters.Add(new SomeCoolNewConverter());
```
without increasing NumberOfConverters to 3, the assert would fail.  And there's no way to write that without having a separate method to include the assert.

YohDeadfall replied:
Missed that. Thanks for pointing out!

YohDeadfall committed:
Inlined array construction of factories

YohDeadfall committed:
Optimized init of default converter lookup

YohDeadfall commented:
@stephentoub, addressed your review. Also I tweaked up initialization of the simple converter dictionary, so it doesn't allocate an enumerable and do not iterates over it.

stephentoub commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:
```diff
@@ -20,41 +20,49 @@ public sealed partial class JsonSerializerOptions
-                converters.Add(converter.TypeToConvert!, converter);
-            }
-
+            // When adding to this, update NumberOfSimpleConverters above.
```
This comment isn't really necessary.  If someone adds to this and doesn't update the number, the assert will fire.

lpereira replied:
You can probably write this as:

```csharp
return new Dictionary<Type, JsonConverter> {
    [JsonConverterBoolean.TypeToConvert!] = new JsonConverterBoolean(),
    [JsonConverterByte.TypeToConvert!] = new JsonConverterByte(),
    [JsonConverterByteArray.TypeToConvert!] = new JsonConverterByteArray(),
    [JsonConverterChar.TypeToConvert!] = new JsonConverterChar(),
    [JsonConverterDateTime.TypeToConvert!] = new JsonConverterDateTime(),
    [JsonConverterDateTimeOffset.TypeToConvert!] = new JsonConverterDateTimeOffset(),
    [JsonConverterDouble.TypeToConvert!] = new JsonConverterDouble(),
    [JsonConverterDecimal.TypeToConvert!] = new JsonConverterDecimal(),
    [JsonConverterGuid.TypeToConvert!] = new JsonConverterGuid(),
    [JsonConverterInt16.TypeToConvert!] = new JsonConverterInt16(),
    [JsonConverterInt32.TypeToConvert!] = new JsonConverterInt32(),
    [JsonConverterInt64.TypeToConvert!] = new JsonConverterInt64(),
    [JsonConverterJsonElement.TypeToConvert!] = new JsonConverterJsonElement(),
    [JsonConverterObject.TypeToConvert!] = new JsonConverterObject(),
    [JsonConverterSByte.TypeToConvert!] = new JsonConverterSByte(),
    [JsonConverterSingle.TypeToConvert!] = new JsonConverterSingle(),
    [JsonConverterString.TypeToConvert!] = new JsonConverterString(),
    [JsonConverterUInt16.TypeToConvert!] = new JsonConverterUInt16(),
    [JsonConverterUInt32.TypeToConvert!] = new JsonConverterUInt32(),
    [JsonConverterUInt64.TypeToConvert!] = new JsonConverterUInt64(),
    [JsonConverterUri.TypeToConvert!] = new JsonConverterUri(),
};
```

It's a little repetitive, yes, but then you get rid of the assertion, NumberOfSimpleConverters, and the comment.

layomia replied:
+1

steveharter replied:
The original reason I added is so you don't have to run a test to figure it out. Anyway it can be removed.

stephentoub commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:
```diff
@@ -20,41 +20,49 @@ public sealed partial class JsonSerializerOptions
-
-            return converters;
+            void Add(JsonConverter converter) =>
+                converters.Add(converter.TypeToConvert!, converter);
```
We'd save 21 virtual calls at start-up if we just spelled these out:
```C#
var converters = new Dictionary<Type, JsonConverter>(NumberOfSimpleConverters)
{
    { typeof(bool), new JsonConverterBoolean() },
    { typeof(byte), new JsonConverterByte() },
    { typeof(byte[]), new new JsonConverterByteArray() },
    ...
};
Debug.Assert(NumberOfSimpleConverters == converters.Count);
return converters;
```

YohDeadfall replied:
Since only `byte[]`, `object`, `string` and `Uri` are classes only for them the runtime will generate a shared generic of type `JsonConverter<>`. For others - type instances will be different, and because all classes are marked as final the JIT is able to inline the code.

Personally I would like to stay with getting a type from a converter for safety and maintainability.

Here is a diff of ASM outputs for [mine](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBY4AcoQGYEsA2MAJgDQAuIAhgM4C2APgAIBMAjALABQjADAASNWKANxcuAYgB2AV3z5KwQnxiSFhMdwDMA5nwBS1CJIDKMKLkr5cALzMB5dGVxHqfAN5c+XvplwA3SjIYAVYkPgARXDAnI0ooAE8AHgAVePQYEn1DSQBhIz8zIKgAPj4AcRgycJhsSlkyY1xadEI8yQKoIuoACgBKT28PTm8RvkhJajI+XEkpgDlpWmB7bEbm1vzCs1cAXj42UWHRrwCoMc3O7b49yRgAdwiomNUElLSMrKM2jqLi7oWlis1i0YN8tlBqL1DgNjgB6WF8ACq1GClD4RCezhe8T42GgfGoTRB53a4OoADoYaN4XwAOoACxUfEoRAxkgA5nwyBAufTcNRMtJ0ERAsEAcsoHZVkSNqTLhDmcAIAVKUdjgBBVndW4PAxfC5FABCEAghEokj6UKpI01RG190+uQNZkN8SClsOxy8tvtuuyYPlrqC6qgUEo8Q91u8Pp1joDRRy9LikbVoxjDr1TrlRXCouSTRgKa9fHTfv12bMuaC+doMCl2BRZCLXtLcedUHCEGkikLvStqZtWtjmfjlZgYCalmbGqHGf97bK0lwdr7npbs7LWZ+ZgAkrNQtO0xu2xWoHuyJpmIfB3bh/PT+fUNfo8eR+3MwBRQi12bP72v+9t0lYAACtxybVco3/W853LIDjCDXt+2LVs31PRoOUIP8SwAuDwWMMhzA5bDUMA8FEXPA9IIHF8YM3UcoAo2ZLxI3Ct3Ix8UFYuiTyAxFzBTKCBAAdhJIDqGhGivEYFAcLtNCgLE8FemuYohJGcZxPJH1NPBclUnSZIIHjABCTJdPlZDvAAXy4WzOC4RhtAUSYw2iHRePBLghhGGYilUfBPPlPp3D4eyqScxVXModylVNPgcnNeNugM4IyHeIz4ysrw/LMAK+D8XBOmkSw+FSgB+MqMuMg1VL4GQ5EOcKtCiwiYqmFggqKFJShALqzG8qlMAgIJomIfqoBCtwwo0EZIuVMxzCIYI4sCxL2M6FL3i5aqsrq9LDJq7Nrj2A6YBwFL+ypXKoHyhbQ2XYJUqqw743294LuSftmpu/KUUscbOoU8FjVNGBzT4PrgflRJVtKNx7N+0r/sIIgPOho03WCKGyJh4AsfhxHZjy5HwdR9Hccx4NQ3DSGJthrGAG0AF1Ca4JHApRwHdAxsxEziOneagRIwCTEp3CJ/zSYBtGgcpys8wLQX5eFqsYBrGA2c4DmCTJ7mJrVjX60bZW8Jhw2C2NyotZ1rnZZ5lXO27JQcbN7qiC7HsbeJ27pfJuW3bHCdaFK12Nvd8dJ3wb2pc5vX7Ymxdl1N8OzESJOiBjkm45linA7PfcwjDhjEkopAs99nP/Yd/Pz0vFOS7r5gK7++O89TguyFQBv21L2ZUBbv39aFz9vxUKZi970eYB/MhB6r4eVbsUDwJ709EggFfonn3Xc4DjuEKxtegMSah8aCHe7fbhiMPZF36dvwhL7b/eb8ImZOUn9fBB4Z+95rjuTEyChGPuCRIQDQh/2rhNIB9cv4n1gc3CW7Mfat3/jAzioCYZAIHsg7WqCh4JyFvxXAWDuokK1o5bQsAWRGHwDidAcQnDI0ItIdy09Z6DU4NZIAA) and [your](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBY4AcoQGYEsA2MAJgDQAuIAhgM4C2APgAIBMAjALABQjADAASNWKANxcuAYgB2AV3z5KwQnxiSFhMdwDMA5nwBS1CJIDKMKLkr5cALzMB5dGVxHqfAN5c+XvplwA3SjIYAVYkPgARXDAnI0ooAE8AHgAVePQYEn1DSQBhIz8zIKgAPj4AcRgycJhsSlkyY1xadEI8yQKoIuoACgBKT28PTm8RvkhJajI+XEkpgDlpWmB7bEbm1vzCs1cAXj42UWHRrwCoMc3O7b49yRgAdwiomNUElLSMrKM2jqLi7oWlis1i0YN8tlBqL1DgNjgB6WF8ACq1GClD4RCezhe8T42GgfGoTRB53a4OoADoYaN4XwAOoACxUfEoRAxkgA5nwyBAufTcNRMtJ0ERAsEAcsoHZVkSNqTLhDmcAIAVKUdjuMfttyQBBVndMjvHDdABCEAghEokl6mVuDwMXwuRVN5pglr6UKpIw1ZJ1eoN6SNxviQWtfFtn1yjrMQaC7sOxy83vlFN1RH1huwJuDMAA2gBdUPh+2RuVO7PaqBQSjxOOe7xJrq+tP+mBGnL0uKF+4RsHy9txWtq0YNrWp9MBzPhUXJJowLt27K9opToIz2gwQcJkmaiFN8etyfT2dS7AosjzntRqArmBrmAns+bhMj3djltG8IQaSKOc27vFpczE/b9CCfdUrxTP0M26aowCaSwLwAq9YPg/AwOHCC93fTMymkXAiEQxcr1w/D0K9TC32ggBJWZQkIh1SzMGiyFCMj6woqCJ26ZjNGYeiSx3HjmDYxMOObajZlQfjAKgZjUBE7cfUorjiwAUUIddZmkq81I0lQyAUl9IPEri7GAAArGBom0xjJQsqyDN6D0h3I2zjP3I1jBjX8w3/IjbK87NDLEjzM0aDlCBsndwvZUCnPjZ8Quw7pjDIcwOSi8FUvS9lgrcrDoMRZi6L/BcGJ3IraKQPKd3c5LKrIXjMvlBreJqpTOIPboGqk0rL1snqUHa5MCq4xFzGaopxtwQc6y8RgAHZFOTBKvAAXy4DbOC4RhtAUSYq2iHR+p3LghhGGYilUfATvBPp3D4LaqV2xUDsoI6lXNPgcktJdulSdIuXeZIICXZyLtmMxrr4PxcE6aRLD4AGYAAfiR4HQcda5ShkORDierRXrS96phYW75RSUoQHJoozqpTAICCaJiBpsx7rcR6NBGF7lTMcwiGCT6bp+gTCn+94gfSEGwexyXb0x0trj2ZLkmcqlLqhxHecrfDgmR9GpYVn5ZZV5yCY1qBoZRSwWbJpDbOdC1JD4an7Z3RIhdKNwtotq3XUIIhjrd8FvJd1moA97MvZ9yHLcR62A6D/yd28isqxxV3k/BSOgnzaOuF9+P/dt3Rg77Dszkz8rs7ACv884QuboTkvw5vO8w7LopEjb2d68bgli8Du2s/lHv1wfSoO5Hrux/vbBT0qPvY79m2h9L6egK/H8p+rimiC3wgl6uovV6T3flys1Cd9FvfL9oSwj81pvB7Pm+ihIwOq7fsxEg/x+4+fqfYe58mJVWvjJRIxUkD/xXonYB39ZKzF4uAq8kCkHMBgSfOB68QGILIKgFBtk0H4JQJgwB2Dw66RgJpKYX8IFUJoWQgeQCcEILMpZI6dDUEQHstEJhzc17h0CkEQh7tqDACju4GOx9yEt07mYGKSguFEMUTAfhL94EyWyjMTkyj3aCB4Oolh4cGqhFEdnUx0CpEF2XlguRG8oCtV0HoixQkjEUPkY4uSKBzEU0Gu4+xuDpq+K7tNeuO1tCwBZEYfAOJ0BxCcPHNK0gjoMP0nTTga0gA) code:

```diff
JsonSerializerOptions.GetDefaultSimpleConverters()
    push rdi
    push rsi
    sub rsp, 0x28
-   mov rcx, 0x7ffa86052fd0
+   mov rcx, 0x7ffa860a2ea8
    call 0x7ffadf4f4690
    mov rsi, rax
    mov rcx, rsi
    mov edx, 0x15
    xor r8d, r8d
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]]..ctor(Int32, System.Collections.Generic.IEqualityComparer`1<System.__Canon>)
-   mov rcx, 0x7ffa8604dfd8
+   mov rcx, 0x7ffa8609dfe8
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f9971c8
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa8604e2c8
+   mov rcx, 0x7ffa8609e2d8
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f999448
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa8604e628
+   mov rcx, 0x7ffa8609e638
    call 0x7ffadf4f4690
    mov rdi, rax
-   mov rcx, rdi
-   call JsonConverter`1[[System.__Canon, System.Private.CoreLib]].get_TypeToConvert()
+   mov rcx, 0x7ffa7fa42522
+   call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa8604e9b0
+   mov rcx, 0x7ffa8609e9c0
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f998000
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa8604eca0
+   mov rcx, 0x7ffa8609ecb0
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7fc265b0
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86050038
+   mov rcx, 0x7ffa860a0038
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7fdf77f0
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86050328
+   mov rcx, 0x7ffa860a0328
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f99e618
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86050618
+   mov rcx, 0x7ffa860a0618
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7fe5a040
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86050908
+   mov rcx, 0x7ffa860a0908
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7fa83d70
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86050bf8
-   call 0x7ffadf4f4690
-   mov rdi, rax
    mov rcx, 0x7ffa7f999e38
    call 0x7ffadf49b730
-   mov rdx, rax
+   mov rdi, rax
+   mov rcx, 0x7ffa860a0bf8
+   call 0x7ffadf4f4690
+   mov r8, rax
    mov rcx, rsi
-   mov r8, rdi
+   mov rdx, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86050ee8
+   mov rcx, 0x7ffa860a0ee8
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f99b1e8
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa860511d8
+   mov rcx, 0x7ffa860a11d8
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f99c568
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa860515e8
+   mov rcx, 0x7ffa860a15e8
    call 0x7ffadf4f4690
    mov rdi, rax
-   mov rcx, 0x7ffa86051318
+   mov rcx, 0x7ffa860a1318
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86051740
+   mov rcx, 0x7ffa860a1740
    call 0x7ffadf4f4690
    mov rdi, rax
-   mov rcx, rdi
-   call JsonConverter`1[[System.__Canon, System.Private.CoreLib]].get_TypeToConvert()
+   mov rcx, 0x7ffa7f990ae8
+   call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86051ac8
+   mov rcx, 0x7ffa860a1ac8
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f998a58
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86051db8
+   mov rcx, 0x7ffa860a1db8
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f99daa0
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86051f10
+   mov rcx, 0x7ffa860a1f10
    call 0x7ffadf4f4690
    mov rdi, rax
-   mov rcx, rdi
-   call JsonConverter`1[[System.__Canon, System.Private.CoreLib]].get_TypeToConvert()
+   mov rcx, 0x7ffa7fa51e18
+   call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86052298
+   mov rcx, 0x7ffa860a2298
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f99a828
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa86052878
+   mov rcx, 0x7ffa860a2588
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f99bba8
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
    mov rcx, 0x7ffa860a2878
    call 0x7ffadf4f4690
    mov rdi, rax
    mov rcx, 0x7ffa7f99cf28
    call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
-   mov rcx, 0x7ffa860529d0
+   mov rcx, 0x7ffa860a29d0
    call 0x7ffadf4f4690
    mov rdi, rax
-   mov rcx, rdi
-   call JsonConverter`1[[System.__Canon, System.Private.CoreLib]].get_TypeToConvert()
+   mov rcx, 0x7ffa7fe3aa50
+   call 0x7ffadf49b730
    mov rdx, rax
    mov rcx, rsi
    mov r8, rdi
    mov r9d, 0x2
    call System.Collections.Generic.Dictionary`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].TryInsert(System.__Canon, System.__Canon, System.Collections.Generic.InsertionBehavior)
    mov rax, rsi
    add rsp, 0x28
    pop rsi
    pop rdi
    ret
```

stephentoub replied:
FWIW, I don't think that assembly output is going to be accurate.  This code gets executed once.  It's either going to be tier 0 / unoptimized, or it's going to be in a R2R image; either way, I'd expect different assembly from what you shared.

It's not a big deal, though.

YohDeadfall replied:
/cc @ahsonkhan @steveharter

layomia replied:
I'd go with the suggestion in https://github.com/dotnet/runtime/pull/2248/files#r371814237, especially with the `DEBUG` validation. I don't think it hurts maintainability.

YohDeadfall replied:
There is a good rule saying not to repeat yourself and specifying a type you're going to duplicate the body of `TypeToConvert`. At the same time code bloats by adding the validation.

Did some benchmarking and here are results. The suffested way (`GetDefaultSimpleConvertersToub`) produces much more code than one with the local `Add` method (`GetDefaultSimpleConvertersYoh`). I do not include results since each method was executed only once, but they had near the same time.

Therefore, I have strong feeling that `Add` method should remain.

---

```
BenchmarkDotNet=v0.12.0, OS=Windows 10.0.19041
Intel Xeon CPU E5-2687W 0 3.10GHz, 2 CPU, 32 logical and 32 physical cores
.NET Core SDK=5.0.100-preview.3.20163.3
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.15501, CoreFX 5.0.20.15501), X64 RyuJIT
  Job-JFQOUO : .NET Core 5.0.0 (CoreCLR 5.0.20.15501, CoreFX 5.0.20.15501), X64 RyuJIT

Runtime=.NET Core 5.0
```

### GetDefaultSimpleConvertersYoh

```
00007ff9`ac83d570 temp.Program.GetDefaultSimpleConvertersYoh()
00007ff9`ac83d5c2 48b9e8bd76acf97f0000 mov rcx,7FF9AC76BDE8h
00007ff9`ac83d5cc e87f67905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d5d1 488945f0        mov     qword ptr [rbp-10h],rax
00007ff9`ac83d5d5 488b4df0        mov     rcx,qword ptr [rbp-10h]
00007ff9`ac83d5d9 ba15000000      mov     edx,15h
00007ff9`ac83d5de e8bda1d9ff      call    00007ff9`ac5d77a0
00007ff9`ac83d5e3 488b4df0        mov     rcx,qword ptr [rbp-10h]
00007ff9`ac83d5e7 48894df8        mov     qword ptr [rbp-8],rcx
00007ff9`ac83d5eb 48b9a0de8bacf97f0000 mov rcx,7FF9AC8BDEA0h
00007ff9`ac83d5f5 e85667905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d5fa 488945e8        mov     qword ptr [rbp-18h],rax
00007ff9`ac83d5fe 488b4de8        mov     rcx,qword ptr [rbp-18h]
00007ff9`ac83d602 e8e1dfffff      call    00007ff9`ac83b5e8
00007ff9`ac83d607 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d60b 488b4de8        mov     rcx,qword ptr [rbp-18h]
00007ff9`ac83d60f e8fc03dbff      call    00007ff9`ac5eda10
00007ff9`ac83d614 48b958e08bacf97f0000 mov rcx,7FF9AC8BE058h
00007ff9`ac83d61e e82d67905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d623 488945e0        mov     qword ptr [rbp-20h],rax
00007ff9`ac83d627 488b4de0        mov     rcx,qword ptr [rbp-20h]
00007ff9`ac83d62b e8c8dfffff      call    00007ff9`ac83b5f8
00007ff9`ac83d630 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d634 488b4de0        mov     rcx,qword ptr [rbp-20h]
00007ff9`ac83d638 e8d303dbff      call    00007ff9`ac5eda10
00007ff9`ac83d63d 48b928e28bacf97f0000 mov rcx,7FF9AC8BE228h
00007ff9`ac83d647 e80467905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d64c 488945d8        mov     qword ptr [rbp-28h],rax
00007ff9`ac83d650 488b4dd8        mov     rcx,qword ptr [rbp-28h]
00007ff9`ac83d654 e8b7edffff      call    00007ff9`ac83c410
00007ff9`ac83d659 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d65d 488b4dd8        mov     rcx,qword ptr [rbp-28h]
00007ff9`ac83d661 e8aa03dbff      call    00007ff9`ac5eda10
00007ff9`ac83d666 48b968e48bacf97f0000 mov rcx,7FF9AC8BE468h
00007ff9`ac83d670 e8db66905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d675 488945d0        mov     qword ptr [rbp-30h],rax
00007ff9`ac83d679 488b4dd0        mov     rcx,qword ptr [rbp-30h]
00007ff9`ac83d67d e8beedffff      call    00007ff9`ac83c440
00007ff9`ac83d682 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d686 488b4dd0        mov     rcx,qword ptr [rbp-30h]
00007ff9`ac83d68a e88103dbff      call    00007ff9`ac5eda10
00007ff9`ac83d68f 48b9c8f98bacf97f0000 mov rcx,7FF9AC8BF9C8h
00007ff9`ac83d699 e8b266905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d69e 488945c8        mov     qword ptr [rbp-38h],rax
00007ff9`ac83d6a2 488b4dc8        mov     rcx,qword ptr [rbp-38h]
00007ff9`ac83d6a6 e885f0ffff      call    00007ff9`ac83c730
00007ff9`ac83d6ab 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d6af 488b4dc8        mov     rcx,qword ptr [rbp-38h]
00007ff9`ac83d6b3 e85803dbff      call    00007ff9`ac5eda10
00007ff9`ac83d6b8 48b9600c8dacf97f0000 mov rcx,7FF9AC8D0C60h
00007ff9`ac83d6c2 e88966905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d6c7 488945c0        mov     qword ptr [rbp-40h],rax
00007ff9`ac83d6cb 488b4dc0        mov     rcx,qword ptr [rbp-40h]
00007ff9`ac83d6cf e8acf0ffff      call    00007ff9`ac83c780
00007ff9`ac83d6d4 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d6d8 488b4dc0        mov     rcx,qword ptr [rbp-40h]
00007ff9`ac83d6dc e82f03dbff      call    00007ff9`ac5eda10
00007ff9`ac83d6e1 48b9780f8dacf97f0000 mov rcx,7FF9AC8D0F78h
00007ff9`ac83d6eb e86066905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d6f0 488945b8        mov     qword ptr [rbp-48h],rax
00007ff9`ac83d6f4 488b4db8        mov     rcx,qword ptr [rbp-48h]
00007ff9`ac83d6f8 e8f3f6ffff      call    00007ff9`ac83cdf0
00007ff9`ac83d6fd 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d701 488b4db8        mov     rcx,qword ptr [rbp-48h]
00007ff9`ac83d705 e80603dbff      call    00007ff9`ac5eda10
00007ff9`ac83d70a 48b930118dacf97f0000 mov rcx,7FF9AC8D1130h
00007ff9`ac83d714 e83766905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d719 488945b0        mov     qword ptr [rbp-50h],rax
00007ff9`ac83d71d 488b4db0        mov     rcx,qword ptr [rbp-50h]
00007ff9`ac83d721 e8faf6ffff      call    00007ff9`ac83ce20
00007ff9`ac83d726 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d72a 488b4db0        mov     rcx,qword ptr [rbp-50h]
00007ff9`ac83d72e e8dd02dbff      call    00007ff9`ac5eda10
00007ff9`ac83d733 48b9e8128dacf97f0000 mov rcx,7FF9AC8D12E8h
00007ff9`ac83d73d e80e66905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d742 488945a8        mov     qword ptr [rbp-58h],rax
00007ff9`ac83d746 488b4da8        mov     rcx,qword ptr [rbp-58h]
00007ff9`ac83d74a e801f7ffff      call    00007ff9`ac83ce50
00007ff9`ac83d74f 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d753 488b4da8        mov     rcx,qword ptr [rbp-58h]
00007ff9`ac83d757 e8b402dbff      call    00007ff9`ac5eda10
00007ff9`ac83d75c 48b9a0148dacf97f0000 mov rcx,7FF9AC8D14A0h
00007ff9`ac83d766 e8e565905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d76b 488945a0        mov     qword ptr [rbp-60h],rax
00007ff9`ac83d76f 488b4da0        mov     rcx,qword ptr [rbp-60h]
00007ff9`ac83d773 e808f7ffff      call    00007ff9`ac83ce80
00007ff9`ac83d778 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d77c 488b4da0        mov     rcx,qword ptr [rbp-60h]
00007ff9`ac83d780 e88b02dbff      call    00007ff9`ac5eda10
00007ff9`ac83d785 48b958168dacf97f0000 mov rcx,7FF9AC8D1658h
00007ff9`ac83d78f e8bc65905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d794 48894598        mov     qword ptr [rbp-68h],rax
00007ff9`ac83d798 488b4d98        mov     rcx,qword ptr [rbp-68h]
00007ff9`ac83d79c e81ff7ffff      call    00007ff9`ac83cec0
00007ff9`ac83d7a1 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d7a5 488b4d98        mov     rcx,qword ptr [rbp-68h]
00007ff9`ac83d7a9 e86202dbff      call    00007ff9`ac5eda10
00007ff9`ac83d7ae 48b910188dacf97f0000 mov rcx,7FF9AC8D1810h
00007ff9`ac83d7b8 e89365905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d7bd 48894590        mov     qword ptr [rbp-70h],rax
00007ff9`ac83d7c1 488b4d90        mov     rcx,qword ptr [rbp-70h]
00007ff9`ac83d7c5 e826f7ffff      call    00007ff9`ac83cef0
00007ff9`ac83d7ca 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d7ce 488b4d90        mov     rcx,qword ptr [rbp-70h]
00007ff9`ac83d7d2 e83902dbff      call    00007ff9`ac5eda10
00007ff9`ac83d7d7 48b9401a8dacf97f0000 mov rcx,7FF9AC8D1A40h
00007ff9`ac83d7e1 e86a65905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d7e6 48894588        mov     qword ptr [rbp-78h],rax
00007ff9`ac83d7ea 488b4d88        mov     rcx,qword ptr [rbp-78h]
00007ff9`ac83d7ee e82df7ffff      call    00007ff9`ac83cf20
00007ff9`ac83d7f3 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d7f7 488b4d88        mov     rcx,qword ptr [rbp-78h]
00007ff9`ac83d7fb e81002dbff      call    00007ff9`ac5eda10
00007ff9`ac83d800 48b9081b8dacf97f0000 mov rcx,7FF9AC8D1B08h
00007ff9`ac83d80a e84165905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d80f 48894580        mov     qword ptr [rbp-80h],rax
00007ff9`ac83d813 488b4d80        mov     rcx,qword ptr [rbp-80h]
00007ff9`ac83d817 e814f7ffff      call    00007ff9`ac83cf30
00007ff9`ac83d81c 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d820 488b4d80        mov     rcx,qword ptr [rbp-80h]
00007ff9`ac83d824 e8e701dbff      call    00007ff9`ac5eda10
00007ff9`ac83d829 48b9481d8dacf97f0000 mov rcx,7FF9AC8D1D48h
00007ff9`ac83d833 e81865905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d838 48898578ffffff  mov     qword ptr [rbp-88h],rax
00007ff9`ac83d83f 488b8d78ffffff  mov     rcx,qword ptr [rbp-88h]
00007ff9`ac83d846 e815f7ffff      call    00007ff9`ac83cf60
00007ff9`ac83d84b 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d84f 488b8d78ffffff  mov     rcx,qword ptr [rbp-88h]
00007ff9`ac83d856 e8b501dbff      call    00007ff9`ac5eda10
00007ff9`ac83d85b 48b9001f8dacf97f0000 mov rcx,7FF9AC8D1F00h
00007ff9`ac83d865 e8e664905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d86a 48898570ffffff  mov     qword ptr [rbp-90h],rax
00007ff9`ac83d871 488b8d70ffffff  mov     rcx,qword ptr [rbp-90h]
00007ff9`ac83d878 e813f7ffff      call    00007ff9`ac83cf90
00007ff9`ac83d87d 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d881 488b8d70ffffff  mov     rcx,qword ptr [rbp-90h]
00007ff9`ac83d888 e88301dbff      call    00007ff9`ac5eda10
00007ff9`ac83d88d 48b9c81f8dacf97f0000 mov rcx,7FF9AC8D1FC8h
00007ff9`ac83d897 e8b464905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d89c 48898568ffffff  mov     qword ptr [rbp-98h],rax
00007ff9`ac83d8a3 488b8d68ffffff  mov     rcx,qword ptr [rbp-98h]
00007ff9`ac83d8aa e8f1f6ffff      call    00007ff9`ac83cfa0
00007ff9`ac83d8af 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d8b3 488b8d68ffffff  mov     rcx,qword ptr [rbp-98h]
00007ff9`ac83d8ba e85101dbff      call    00007ff9`ac5eda10
00007ff9`ac83d8bf 48b908228dacf97f0000 mov rcx,7FF9AC8D2208h
00007ff9`ac83d8c9 e88264905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d8ce 48898560ffffff  mov     qword ptr [rbp-0A0h],rax
00007ff9`ac83d8d5 488b8d60ffffff  mov     rcx,qword ptr [rbp-0A0h]
00007ff9`ac83d8dc e8eff6ffff      call    00007ff9`ac83cfd0
00007ff9`ac83d8e1 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d8e5 488b8d60ffffff  mov     rcx,qword ptr [rbp-0A0h]
00007ff9`ac83d8ec e81f01dbff      call    00007ff9`ac5eda10
00007ff9`ac83d8f1 48b9c0238dacf97f0000 mov rcx,7FF9AC8D23C0h
00007ff9`ac83d8fb e85064905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d900 48898558ffffff  mov     qword ptr [rbp-0A8h],rax
00007ff9`ac83d907 488b8d58ffffff  mov     rcx,qword ptr [rbp-0A8h]
00007ff9`ac83d90e e8edf6ffff      call    00007ff9`ac83d000
00007ff9`ac83d913 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d917 488b8d58ffffff  mov     rcx,qword ptr [rbp-0A8h]
00007ff9`ac83d91e e8ed00dbff      call    00007ff9`ac5eda10
00007ff9`ac83d923 48b978258dacf97f0000 mov rcx,7FF9AC8D2578h
00007ff9`ac83d92d e81e64905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d932 48898550ffffff  mov     qword ptr [rbp-0B0h],rax
00007ff9`ac83d939 488b8d50ffffff  mov     rcx,qword ptr [rbp-0B0h]
00007ff9`ac83d940 e8ebf6ffff      call    00007ff9`ac83d030
00007ff9`ac83d945 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d949 488b8d50ffffff  mov     rcx,qword ptr [rbp-0B0h]
00007ff9`ac83d950 e8bb00dbff      call    00007ff9`ac5eda10
00007ff9`ac83d955 48b940268dacf97f0000 mov rcx,7FF9AC8D2640h
00007ff9`ac83d95f e8ec63905f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac83d964 48898548ffffff  mov     qword ptr [rbp-0B8h],rax
00007ff9`ac83d96b 488b8d48ffffff  mov     rcx,qword ptr [rbp-0B8h]
00007ff9`ac83d972 e8c9f6ffff      call    00007ff9`ac83d040
00007ff9`ac83d977 488d55f8        lea     rdx,[rbp-8]
00007ff9`ac83d97b 488b8d48ffffff  mov     rcx,qword ptr [rbp-0B8h]
00007ff9`ac83d982 e88900dbff      call    00007ff9`ac5eda10
00007ff9`ac83d987 488b45f8        mov     rax,qword ptr [rbp-8]

00007ff9`ac83da70 temp.Program.g__Add|1_0(temp.JsonConverter, <>c__DisplayClass1_0 ByRef)
00007ff9`ac83da8c 488b4d18        mov     rcx,qword ptr [rbp+18h]
00007ff9`ac83da90 488b09          mov     rcx,qword ptr [rcx]
00007ff9`ac83da93 48894df8        mov     qword ptr [rbp-8],rcx
00007ff9`ac83da97 488b4d10        mov     rcx,qword ptr [rbp+10h]
00007ff9`ac83da9b 488b4510        mov     rax,qword ptr [rbp+10h]
00007ff9`ac83da9f 488b00          mov     rax,qword ptr [rax]
00007ff9`ac83daa2 488b4040        mov     rax,qword ptr [rax+40h]
00007ff9`ac83daa6 ff5028          call    qword ptr [rax+28h]
00007ff9`ac83daa9 488945f0        mov     qword ptr [rbp-10h],rax
00007ff9`ac83daad 488b55f0        mov     rdx,qword ptr [rbp-10h]
00007ff9`ac83dab1 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac83dab5 4c8b4510        mov     r8,qword ptr [rbp+10h]
00007ff9`ac83dab9 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac83dabb ff157754e3ff    call    qword ptr [00007ff9`ac672f38]
00007ff9`ac83dac1 90              nop
```


### GetDefaultSimpleConvertersToub

```
00007ff9`ac84d570 temp.Program.GetDefaultSimpleConvertersToub()
00007ff9`ac84d5b9 48b9e8bd77acf97f0000 mov rcx,7FF9AC77BDE8h
00007ff9`ac84d5c3 e888678f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d5c8 488945f8        mov     qword ptr [rbp-8],rax
00007ff9`ac84d5cc 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d5d0 ba15000000      mov     edx,15h
00007ff9`ac84d5d5 e8c6a1d9ff      call    00007ff9`ac5e77a0
00007ff9`ac84d5da 48b9b8b75bacf97f0000 mov rcx,7FF9AC5BB7B8h
00007ff9`ac84d5e4 e8a7f2865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d5e9 488945f0        mov     qword ptr [rbp-10h],rax
00007ff9`ac84d5ed 48b900de8cacf97f0000 mov rcx,7FF9AC8CDE00h
00007ff9`ac84d5f7 e854678f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d5fc 488945e8        mov     qword ptr [rbp-18h],rax
00007ff9`ac84d600 488b4df0        mov     rcx,qword ptr [rbp-10h]
00007ff9`ac84d604 e8274f8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d609 488945e0        mov     qword ptr [rbp-20h],rax
00007ff9`ac84d60d 488b4de8        mov     rcx,qword ptr [rbp-18h]
00007ff9`ac84d611 e8d2dfffff      call    00007ff9`ac84b5e8
00007ff9`ac84d616 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d61a 488b55e0        mov     rdx,qword ptr [rbp-20h]
00007ff9`ac84d61e 4c8b45e8        mov     r8,qword ptr [rbp-18h]
00007ff9`ac84d622 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d624 ff150e59e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d62a 48b938da5bacf97f0000 mov rcx,7FF9AC5BDA38h
00007ff9`ac84d634 e857f2865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d639 488945d8        mov     qword ptr [rbp-28h],rax
00007ff9`ac84d63d 48b9b8df8cacf97f0000 mov rcx,7FF9AC8CDFB8h
00007ff9`ac84d647 e804678f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d64c 488945d0        mov     qword ptr [rbp-30h],rax
00007ff9`ac84d650 488b4dd8        mov     rcx,qword ptr [rbp-28h]
00007ff9`ac84d654 e8d74e8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d659 488945c8        mov     qword ptr [rbp-38h],rax
00007ff9`ac84d65d 488b4dd0        mov     rcx,qword ptr [rbp-30h]
00007ff9`ac84d661 e892dfffff      call    00007ff9`ac84b5f8
00007ff9`ac84d666 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d66a 488b55c8        mov     rdx,qword ptr [rbp-38h]
00007ff9`ac84d66e 4c8b45d0        mov     r8,qword ptr [rbp-30h]
00007ff9`ac84d672 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d674 ff15be58e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d67a 48b9a0f469acf97f0000 mov rcx,7FF9AC69F4A0h
00007ff9`ac84d684 e807f2865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d689 488945c0        mov     qword ptr [rbp-40h],rax
00007ff9`ac84d68d 48b988e18cacf97f0000 mov rcx,7FF9AC8CE188h
00007ff9`ac84d697 e8b4668f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d69c 488945b8        mov     qword ptr [rbp-48h],rax
00007ff9`ac84d6a0 488b4dc0        mov     rcx,qword ptr [rbp-40h]
00007ff9`ac84d6a4 e8874e8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d6a9 488945b0        mov     qword ptr [rbp-50h],rax
00007ff9`ac84d6ad 488b4db8        mov     rcx,qword ptr [rbp-48h]
00007ff9`ac84d6b1 e85aedffff      call    00007ff9`ac84c410
00007ff9`ac84d6b6 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d6ba 488b55b0        mov     rdx,qword ptr [rbp-50h]
00007ff9`ac84d6be 4c8b45b8        mov     r8,qword ptr [rbp-48h]
00007ff9`ac84d6c2 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d6c4 ff156e58e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d6ca 48b9f0c55bacf97f0000 mov rcx,7FF9AC5BC5F0h
00007ff9`ac84d6d4 e8b7f1865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d6d9 488945a8        mov     qword ptr [rbp-58h],rax
00007ff9`ac84d6dd 48b9c8e38cacf97f0000 mov rcx,7FF9AC8CE3C8h
00007ff9`ac84d6e7 e864668f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d6ec 488945a0        mov     qword ptr [rbp-60h],rax
00007ff9`ac84d6f0 488b4da8        mov     rcx,qword ptr [rbp-58h]
00007ff9`ac84d6f4 e8374e8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d6f9 48894598        mov     qword ptr [rbp-68h],rax
00007ff9`ac84d6fd 488b4da0        mov     rcx,qword ptr [rbp-60h]
00007ff9`ac84d701 e83aedffff      call    00007ff9`ac84c440
00007ff9`ac84d706 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d70a 488b5598        mov     rdx,qword ptr [rbp-68h]
00007ff9`ac84d70e 4c8b45a0        mov     r8,qword ptr [rbp-60h]
00007ff9`ac84d712 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d714 ff151e58e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d71a 48b978f68cacf97f0000 mov rcx,7FF9AC8CF678h
00007ff9`ac84d724 e867f1865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d729 48894590        mov     qword ptr [rbp-70h],rax
00007ff9`ac84d72d 48b988fa8cacf97f0000 mov rcx,7FF9AC8CFA88h
00007ff9`ac84d737 e814668f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d73c 48894588        mov     qword ptr [rbp-78h],rax
00007ff9`ac84d740 488b4d90        mov     rcx,qword ptr [rbp-70h]
00007ff9`ac84d744 e8e74d8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d749 48894580        mov     qword ptr [rbp-80h],rax
00007ff9`ac84d74d 488b4d88        mov     rcx,qword ptr [rbp-78h]
00007ff9`ac84d751 e8faefffff      call    00007ff9`ac84c750
00007ff9`ac84d756 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d75a 488b5580        mov     rdx,qword ptr [rbp-80h]
00007ff9`ac84d75e 4c8b4588        mov     r8,qword ptr [rbp-78h]
00007ff9`ac84d762 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d764 ff15ce57e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d76a 48b9400a8eacf97f0000 mov rcx,7FF9AC8E0A40h
00007ff9`ac84d774 e817f1865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d779 48898578ffffff  mov     qword ptr [rbp-88h],rax
00007ff9`ac84d780 48b9c00d8eacf97f0000 mov rcx,7FF9AC8E0DC0h
00007ff9`ac84d78a e8c1658f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d78f 48898570ffffff  mov     qword ptr [rbp-90h],rax
00007ff9`ac84d796 488b8d78ffffff  mov     rcx,qword ptr [rbp-88h]
00007ff9`ac84d79d e88e4d8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d7a2 48898568ffffff  mov     qword ptr [rbp-98h],rax
00007ff9`ac84d7a9 488b8d70ffffff  mov     rcx,qword ptr [rbp-90h]
00007ff9`ac84d7b0 e8ebefffff      call    00007ff9`ac84c7a0
00007ff9`ac84d7b5 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d7b9 488b9568ffffff  mov     rdx,qword ptr [rbp-98h]
00007ff9`ac84d7c0 4c8b8570ffffff  mov     r8,qword ptr [rbp-90h]
00007ff9`ac84d7c7 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d7c9 ff156957e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d7cf 48b9683167acf97f0000 mov rcx,7FF9AC673168h
00007ff9`ac84d7d9 e8b2f0865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d7de 48898560ffffff  mov     qword ptr [rbp-0A0h],rax
00007ff9`ac84d7e5 48b9780f8eacf97f0000 mov rcx,7FF9AC8E0F78h
00007ff9`ac84d7ef e85c658f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d7f4 48898558ffffff  mov     qword ptr [rbp-0A8h],rax
00007ff9`ac84d7fb 488b8d60ffffff  mov     rcx,qword ptr [rbp-0A0h]
00007ff9`ac84d802 e8294d8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d807 48898550ffffff  mov     qword ptr [rbp-0B0h],rax
00007ff9`ac84d80e 488b8d58ffffff  mov     rcx,qword ptr [rbp-0A8h]
00007ff9`ac84d815 e8d6f5ffff      call    00007ff9`ac84cdf0
00007ff9`ac84d81a 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d81e 488b9550ffffff  mov     rdx,qword ptr [rbp-0B0h]
00007ff9`ac84d825 4c8b8558ffffff  mov     r8,qword ptr [rbp-0A8h]
00007ff9`ac84d82c 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d82e ff150457e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d834 48b9285067acf97f0000 mov rcx,7FF9AC675028h
00007ff9`ac84d83e e84df0865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d843 48898548ffffff  mov     qword ptr [rbp-0B8h],rax
00007ff9`ac84d84a 48b930118eacf97f0000 mov rcx,7FF9AC8E1130h
00007ff9`ac84d854 e8f7648f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d859 48898540ffffff  mov     qword ptr [rbp-0C0h],rax
00007ff9`ac84d860 488b8d48ffffff  mov     rcx,qword ptr [rbp-0B8h]
00007ff9`ac84d867 e8c44c8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d86c 48898538ffffff  mov     qword ptr [rbp-0C8h],rax
00007ff9`ac84d873 488b8d40ffffff  mov     rcx,qword ptr [rbp-0C0h]
00007ff9`ac84d87a e8a1f5ffff      call    00007ff9`ac84ce20
00007ff9`ac84d87f 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d883 488b9538ffffff  mov     rdx,qword ptr [rbp-0C8h]
00007ff9`ac84d88a 4c8b8540ffffff  mov     r8,qword ptr [rbp-0C0h]
00007ff9`ac84d891 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d893 ff159f56e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d899 48b9a08169acf97f0000 mov rcx,7FF9AC6981A0h
00007ff9`ac84d8a3 e8e8ef865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d8a8 48898530ffffff  mov     qword ptr [rbp-0D0h],rax
00007ff9`ac84d8af 48b9e8128eacf97f0000 mov rcx,7FF9AC8E12E8h
00007ff9`ac84d8b9 e892648f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d8be 48898528ffffff  mov     qword ptr [rbp-0D8h],rax
00007ff9`ac84d8c5 488b8d30ffffff  mov     rcx,qword ptr [rbp-0D0h]
00007ff9`ac84d8cc e85f4c8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d8d1 48898520ffffff  mov     qword ptr [rbp-0E0h],rax
00007ff9`ac84d8d8 488b8d28ffffff  mov     rcx,qword ptr [rbp-0D8h]
00007ff9`ac84d8df e86cf5ffff      call    00007ff9`ac84ce50
00007ff9`ac84d8e4 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d8e8 488b9520ffffff  mov     rdx,qword ptr [rbp-0E0h]
00007ff9`ac84d8ef 4c8b8528ffffff  mov     r8,qword ptr [rbp-0D8h]
00007ff9`ac84d8f6 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d8f8 ff153a56e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d8fe 48b928e45bacf97f0000 mov rcx,7FF9AC5BE428h
00007ff9`ac84d908 e883ef865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d90d 48898518ffffff  mov     qword ptr [rbp-0E8h],rax
00007ff9`ac84d914 488b8d18ffffff  mov     rcx,qword ptr [rbp-0E8h]
00007ff9`ac84d91b e8104c8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d920 48898510ffffff  mov     qword ptr [rbp-0F0h],rax
00007ff9`ac84d927 48b9a0148eacf97f0000 mov rcx,7FF9AC8E14A0h
00007ff9`ac84d931 e81a648f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d936 48898508ffffff  mov     qword ptr [rbp-0F8h],rax
00007ff9`ac84d93d 488b8d08ffffff  mov     rcx,qword ptr [rbp-0F8h]
00007ff9`ac84d944 e837f5ffff      call    00007ff9`ac84ce80
00007ff9`ac84d949 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d94d 488b9510ffffff  mov     rdx,qword ptr [rbp-0F0h]
00007ff9`ac84d954 4c8b8508ffffff  mov     r8,qword ptr [rbp-0F8h]
00007ff9`ac84d95b 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d95d ff15d555e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d963 48b9d8f75bacf97f0000 mov rcx,7FF9AC5BF7D8h
00007ff9`ac84d96d e81eef865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d972 48898500ffffff  mov     qword ptr [rbp-100h],rax
00007ff9`ac84d979 48b958168eacf97f0000 mov rcx,7FF9AC8E1658h
00007ff9`ac84d983 e8c8638f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d988 488985f8feffff  mov     qword ptr [rbp-108h],rax
00007ff9`ac84d98f 488b8d00ffffff  mov     rcx,qword ptr [rbp-100h]
00007ff9`ac84d996 e8954b8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84d99b 488985f0feffff  mov     qword ptr [rbp-110h],rax
00007ff9`ac84d9a2 488b8df8feffff  mov     rcx,qword ptr [rbp-108h]
00007ff9`ac84d9a9 e812f5ffff      call    00007ff9`ac84cec0
00007ff9`ac84d9ae 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84d9b2 488b95f0feffff  mov     rdx,qword ptr [rbp-110h]
00007ff9`ac84d9b9 4c8b85f8feffff  mov     r8,qword ptr [rbp-108h]
00007ff9`ac84d9c0 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84d9c2 ff157055e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84d9c8 48b9b81067acf97f0000 mov rcx,7FF9AC6710B8h
00007ff9`ac84d9d2 e8b9ee865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84d9d7 488985e8feffff  mov     qword ptr [rbp-118h],rax
00007ff9`ac84d9de 48b910188eacf97f0000 mov rcx,7FF9AC8E1810h
00007ff9`ac84d9e8 e863638f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84d9ed 488985e0feffff  mov     qword ptr [rbp-120h],rax
00007ff9`ac84d9f4 488b8de8feffff  mov     rcx,qword ptr [rbp-118h]
00007ff9`ac84d9fb e8304b8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84da00 488985d8feffff  mov     qword ptr [rbp-128h],rax
00007ff9`ac84da07 488b8de0feffff  mov     rcx,qword ptr [rbp-120h]
00007ff9`ac84da0e e8ddf4ffff      call    00007ff9`ac84cef0
00007ff9`ac84da13 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84da17 488b95d8feffff  mov     rdx,qword ptr [rbp-128h]
00007ff9`ac84da1e 4c8b85e0feffff  mov     r8,qword ptr [rbp-120h]
00007ff9`ac84da25 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84da27 ff150b55e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84da2d 48b9a8188eacf97f0000 mov rcx,7FF9AC8E18A8h
00007ff9`ac84da37 e854ee865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84da3c 488985d0feffff  mov     qword ptr [rbp-130h],rax
00007ff9`ac84da43 48b9401a8eacf97f0000 mov rcx,7FF9AC8E1A40h
00007ff9`ac84da4d e8fe628f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84da52 488985c8feffff  mov     qword ptr [rbp-138h],rax
00007ff9`ac84da59 488b8dd0feffff  mov     rcx,qword ptr [rbp-130h]
00007ff9`ac84da60 e8cb4a8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84da65 488985c0feffff  mov     qword ptr [rbp-140h],rax
00007ff9`ac84da6c 488b8dc8feffff  mov     rcx,qword ptr [rbp-138h]
00007ff9`ac84da73 e8a8f4ffff      call    00007ff9`ac84cf20
00007ff9`ac84da78 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84da7c 488b95c0feffff  mov     rdx,qword ptr [rbp-140h]
00007ff9`ac84da83 4c8b85c8feffff  mov     r8,qword ptr [rbp-138h]
00007ff9`ac84da8a 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84da8c ff15a654e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84da92 48b950525bacf97f0000 mov rcx,7FF9AC5B5250h
00007ff9`ac84da9c e8efed865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84daa1 488985b8feffff  mov     qword ptr [rbp-148h],rax
00007ff9`ac84daa8 48b9081b8eacf97f0000 mov rcx,7FF9AC8E1B08h
00007ff9`ac84dab2 e899628f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84dab7 488985b0feffff  mov     qword ptr [rbp-150h],rax
00007ff9`ac84dabe 488b8db8feffff  mov     rcx,qword ptr [rbp-148h]
00007ff9`ac84dac5 e8664a8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84daca 488985a8feffff  mov     qword ptr [rbp-158h],rax
00007ff9`ac84dad1 488b8db0feffff  mov     rcx,qword ptr [rbp-150h]
00007ff9`ac84dad8 e853f4ffff      call    00007ff9`ac84cf30
00007ff9`ac84dadd 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84dae1 488b95a8feffff  mov     rdx,qword ptr [rbp-158h]
00007ff9`ac84dae8 4c8b85b0feffff  mov     r8,qword ptr [rbp-150h]
00007ff9`ac84daef 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84daf1 ff154154e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84daf7 48b948d05bacf97f0000 mov rcx,7FF9AC5BD048h
00007ff9`ac84db01 e88aed865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84db06 488985a0feffff  mov     qword ptr [rbp-160h],rax
00007ff9`ac84db0d 48b9481d8eacf97f0000 mov rcx,7FF9AC8E1D48h
00007ff9`ac84db17 e834628f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84db1c 48898598feffff  mov     qword ptr [rbp-168h],rax
00007ff9`ac84db23 488b8da0feffff  mov     rcx,qword ptr [rbp-160h]
00007ff9`ac84db2a e8014a8b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84db2f 48898590feffff  mov     qword ptr [rbp-170h],rax
00007ff9`ac84db36 488b8d98feffff  mov     rcx,qword ptr [rbp-168h]
00007ff9`ac84db3d e81ef4ffff      call    00007ff9`ac84cf60
00007ff9`ac84db42 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84db46 488b9590feffff  mov     rdx,qword ptr [rbp-170h]
00007ff9`ac84db4d 4c8b8598feffff  mov     r8,qword ptr [rbp-168h]
00007ff9`ac84db54 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84db56 ff15dc53e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84db5c 48b9f02567acf97f0000 mov rcx,7FF9AC6725F0h
00007ff9`ac84db66 e825ed865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84db6b 48898588feffff  mov     qword ptr [rbp-178h],rax
00007ff9`ac84db72 48b9001f8eacf97f0000 mov rcx,7FF9AC8E1F00h
00007ff9`ac84db7c e8cf618f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84db81 48898580feffff  mov     qword ptr [rbp-180h],rax
00007ff9`ac84db88 488b8d88feffff  mov     rcx,qword ptr [rbp-178h]
00007ff9`ac84db8f e89c498b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84db94 48898578feffff  mov     qword ptr [rbp-188h],rax
00007ff9`ac84db9b 488b8d80feffff  mov     rcx,qword ptr [rbp-180h]
00007ff9`ac84dba2 e8e9f3ffff      call    00007ff9`ac84cf90
00007ff9`ac84dba7 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84dbab 488b9578feffff  mov     rdx,qword ptr [rbp-188h]
00007ff9`ac84dbb2 4c8b8580feffff  mov     r8,qword ptr [rbp-180h]
00007ff9`ac84dbb9 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84dbbb ff157753e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84dbc1 48b9b8b667acf97f0000 mov rcx,7FF9AC67B6B8h
00007ff9`ac84dbcb e8c0ec865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84dbd0 48898570feffff  mov     qword ptr [rbp-190h],rax
00007ff9`ac84dbd7 48b9c81f8eacf97f0000 mov rcx,7FF9AC8E1FC8h
00007ff9`ac84dbe1 e86a618f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84dbe6 48898568feffff  mov     qword ptr [rbp-198h],rax
00007ff9`ac84dbed 488b8d70feffff  mov     rcx,qword ptr [rbp-190h]
00007ff9`ac84dbf4 e837498b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84dbf9 48898560feffff  mov     qword ptr [rbp-1A0h],rax
00007ff9`ac84dc00 488b8d68feffff  mov     rcx,qword ptr [rbp-198h]
00007ff9`ac84dc07 e894f3ffff      call    00007ff9`ac84cfa0
00007ff9`ac84dc0c 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84dc10 488b9560feffff  mov     rdx,qword ptr [rbp-1A0h]
00007ff9`ac84dc17 4c8b8568feffff  mov     r8,qword ptr [rbp-198h]
00007ff9`ac84dc1e 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84dc20 ff151253e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84dc26 48b918ee5bacf97f0000 mov rcx,7FF9AC5BEE18h
00007ff9`ac84dc30 e85bec865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84dc35 48898558feffff  mov     qword ptr [rbp-1A8h],rax
00007ff9`ac84dc3c 48b908228eacf97f0000 mov rcx,7FF9AC8E2208h
00007ff9`ac84dc46 e805618f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84dc4b 48898550feffff  mov     qword ptr [rbp-1B0h],rax
00007ff9`ac84dc52 488b8d58feffff  mov     rcx,qword ptr [rbp-1A8h]
00007ff9`ac84dc59 e8d2488b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84dc5e 48898548feffff  mov     qword ptr [rbp-1B8h],rax
00007ff9`ac84dc65 488b8d50feffff  mov     rcx,qword ptr [rbp-1B0h]
00007ff9`ac84dc6c e85ff3ffff      call    00007ff9`ac84cfd0
00007ff9`ac84dc71 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84dc75 488b9548feffff  mov     rdx,qword ptr [rbp-1B8h]
00007ff9`ac84dc7c 4c8b8550feffff  mov     r8,qword ptr [rbp-1B0h]
00007ff9`ac84dc83 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84dc85 ff15ad52e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84dc8b 48b9f80667acf97f0000 mov rcx,7FF9AC6706F8h
00007ff9`ac84dc95 e8f6eb865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84dc9a 48898540feffff  mov     qword ptr [rbp-1C0h],rax
00007ff9`ac84dca1 48b9c0238eacf97f0000 mov rcx,7FF9AC8E23C0h
00007ff9`ac84dcab e8a0608f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84dcb0 48898538feffff  mov     qword ptr [rbp-1C8h],rax
00007ff9`ac84dcb7 488b8d40feffff  mov     rcx,qword ptr [rbp-1C0h]
00007ff9`ac84dcbe e86d488b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84dcc3 48898530feffff  mov     qword ptr [rbp-1D0h],rax
00007ff9`ac84dcca 488b8d38feffff  mov     rcx,qword ptr [rbp-1C8h]
00007ff9`ac84dcd1 e82af3ffff      call    00007ff9`ac84d000
00007ff9`ac84dcd6 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84dcda 488b9530feffff  mov     rdx,qword ptr [rbp-1D0h]
00007ff9`ac84dce1 4c8b8538feffff  mov     r8,qword ptr [rbp-1C8h]
00007ff9`ac84dce8 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84dcea ff154852e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84dcf0 48b9781a67acf97f0000 mov rcx,7FF9AC671A78h
00007ff9`ac84dcfa e891eb865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84dcff 48898528feffff  mov     qword ptr [rbp-1D8h],rax
00007ff9`ac84dd06 48b978258eacf97f0000 mov rcx,7FF9AC8E2578h
00007ff9`ac84dd10 e83b608f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84dd15 48898520feffff  mov     qword ptr [rbp-1E0h],rax
00007ff9`ac84dd1c 488b8d28feffff  mov     rcx,qword ptr [rbp-1D8h]
00007ff9`ac84dd23 e808488b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84dd28 48898518feffff  mov     qword ptr [rbp-1E8h],rax
00007ff9`ac84dd2f 488b8d20feffff  mov     rcx,qword ptr [rbp-1E0h]
00007ff9`ac84dd36 e8f5f2ffff      call    00007ff9`ac84d030
00007ff9`ac84dd3b 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84dd3f 488b9518feffff  mov     rdx,qword ptr [rbp-1E8h]
00007ff9`ac84dd46 4c8b8520feffff  mov     r8,qword ptr [rbp-1E0h]
00007ff9`ac84dd4d 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84dd4f ff15e351e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84dd55 48b9c03e8eacf97f0000 mov rcx,7FF9AC8E3EC0h
00007ff9`ac84dd5f e82ceb865f      call    coreclr+0xbc890 (00007ffa`0c0bc890)
00007ff9`ac84dd64 48898510feffff  mov     qword ptr [rbp-1F0h],rax
00007ff9`ac84dd6b 48b9c83f8eacf97f0000 mov rcx,7FF9AC8E3FC8h
00007ff9`ac84dd75 e8d65f8f5f      call    coreclr!coreclr_shutdown_2+0xfd80 (00007ffa`0c143d50)
00007ff9`ac84dd7a 48898508feffff  mov     qword ptr [rbp-1F8h],rax
00007ff9`ac84dd81 488b8d10feffff  mov     rcx,qword ptr [rbp-1F0h]
00007ff9`ac84dd88 e8a3478b5f      call    coreclr+0x102530 (00007ffa`0c102530)
00007ff9`ac84dd8d 48898500feffff  mov     qword ptr [rbp-200h],rax
00007ff9`ac84dd94 488b8d08feffff  mov     rcx,qword ptr [rbp-1F8h]
00007ff9`ac84dd9b e850f4ffff      call    00007ff9`ac84d1f0
00007ff9`ac84dda0 488b4df8        mov     rcx,qword ptr [rbp-8]
00007ff9`ac84dda4 488b9500feffff  mov     rdx,qword ptr [rbp-200h]
00007ff9`ac84ddab 4c8b8508feffff  mov     r8,qword ptr [rbp-1F8h]
00007ff9`ac84ddb2 3909            cmp     dword ptr [rcx],ecx
00007ff9`ac84ddb4 ff157e51e3ff    call    qword ptr [00007ff9`ac682f38]
00007ff9`ac84ddba 488b45f8        mov     rax,qword ptr [rbp-8]
```

steveharter replied:
>> Therefore, I have strong feeling that Add method should remain.

I prefer the Add method. It is more readable\understandable and ensures TypeToConvert is correct. This code only runs once so I'm not concerned about perf here although removing the yield statements makes sense since they are unnecessary.

layomia commented:
@YohDeadfall we can move forward with this PR when it is synced with master.

YohDeadfall commented:
@layomia, done.

stephentoub commented:
Sorry, @YohDeadfall, looks like it needs to be rebased again.

stephentoub commented:
@layomia, @steveharter, @ahsonkhan, once this is rebased, are you happy with it, or is there more review you want to do?  Can you please stay on top of it and merge it when it's ready?

layomia commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:
```diff
@@ -20,48 +20,56 @@ public sealed partial class JsonSerializerOptions
+            new NullableConverterFactory(),
+            new EnumConverterFactory(),
+            new KeyValuePairConverterFactory(),
+            // IEnumerable should always be last since they can convert any IEnumerable.
```
Please remove as `IEnumerable` isn't the last.

layomia reviewed (commented):
Some comments. Otherwise, LGTM pending rebase and CI.

YohDeadfall committed:
Removed comments

steveharter commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:
```diff
@@ -26,9 +26,7 @@ public sealed partial class JsonSerializerOptions
             new NullableConverterFactory(),
             new EnumConverterFactory(),
             new KeyValuePairConverterFactory(),
-            // IEnumerable should always be last since they can convert any IEnumerable.
```
IEnumerable should come second to last. This comment is valid otherwise

steveharter commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:
```diff
@@ -26,9 +26,7 @@ public sealed partial class JsonSerializerOptions
             new KeyValuePairConverterFactory(),
-            // IEnumerable should always be last since they can convert any IEnumerable.
             new IEnumerableConverterFactory(),
-            // Object should always be last since it converts any type.
```
This comment is valid

steveharter commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:
```diff
@@ -20,48 +20,54 @@ public sealed partial class JsonSerializerOptions
+            new NullableConverterFactory(),
+            new EnumConverterFactory(),
+            new KeyValuePairConverterFactory(),
+            new IEnumerableConverterFactory(),
```
Please add comments back here - IEnumerableConverterFactory should always be second to last and ObjectConverterFactory should always be last.

steveharter commented:
LGTM if the two comments are added back (for object and IEnumerable factories). I realize the IEnumerable comment was incorrect previously (should be second to last, not last).

YohDeadfall committed:
Returned comments back

YohDeadfall commented:
@steveharter, added comments back as requested.

layomia reviewed (approved):
Thanks.
